adding Model Flattening to RFC (#10092)
Azure/azure-powershell: documentation/RFC/Model Flattening and Inline Creation .md

RFC: Parameter Flattening and Inline Creation
Author: Scott Phibbs
Status:
SupersedeBy:
Area:
---

# Flattening to simple parameter types for easy in-line creation
To improve consistency and simplify creation scenarios, users will no longer need to build configurations or complex types to create Azure Resources. Complex types will be flattened into multiple parameters of simple types. Return types will remain the same.

## Motivation

> More consistency across Azure PowerShell commands is well known user feedback. While both flattening and Inline Creation improvements address the consistency concerns, let's focus more on the specific changes. Creating complex types prior to calling creation commands, whether that be by creating your own hashtable or calling a cmdlet to create an in memory complex object is a setup step that will no longer be necessary. We have done this by flattening complex types into simple type parameters. Users will be able to directly call a new or update command in a single call now. To demonstrate these changes we will look at the generic resource creation command `New-AzResource` for flattened parameters and a Networking scenario that no longer needs pre created in-memory configs.

Reviewer: ```suggestion
More consistency across Azure PowerShell commands is well known user feedback. While both flattening and Inline Creation improvements address the consistency concerns, let's focus more on the specific changes. Creating complex types prior to calling creation commands, whether that is by creating your own hashtable or calling a cmdlet to create an in memory complex object is a setup step that will no longer be necessary. We have made this possible by flattening complex types into simple type parameters. Using the Az 4.0 preview, you will be able to create or modify a resource using a single cmdlet execution. To demonstrate these changes, we will look at the generic resource creation command `New-AzResource` for flattened parameters and a Networking scenario that no longer needs pre created in-memory configurations.
```

## User Experience

> `New-AzResource` is an example of where we have flattened `Hashtable` parameters into simple types. The `-Plan` and `-Sku` parameters are both examples of parameters that previously accepted a hashtable and have now been flattened. First, let's look at the parameters for `New-AzResource` in `Az` 2.6. I only show the `[-Plan <Hashtable>]` and `[-Sku <Hashtable>]` specifically from the help content.

Reviewer: ```suggestion
`New-AzResource` is an example of where we have flattened `Hashtable` parameters into simple types. The `-Plan` and `-Sku` parameters are both examples of parameters that previously accepted a hashtable and have now been flattened. First, let's look at the `Plan` and `Sku` parameters for `New-AzResource` in `Az` 2.6.
```

New-AzResource `
    -Name storageskuhash `
    -ResourceGroupName skurg `
    -ResourceType "Microsoft.Storage/storageAccounts" `
    -Sku $sku `
    -Kind "Storage" `
    -ApiVersion "2019-04-01" `
    -Location centralus
```

In `Az` 4.0 (Preview), we have flattened both `-Sku` and `-Plan` to the individual simple types expected in the hashtable. Here I have stripped down the help content to only show the new, flattened 'Sku' and 'Plan' parameters:

```
New-AzResource
    ...
    [-PlanName <String>]
    [-PlanProduct <String>]
    [-PlanPromotionCode <String>]
    [-PlanPublisher <String>]
    [-PlanVersion <String>]
    [-SkuCapacity <Int32>]
    [-SkuFamily <String>]
    [-SkuModel <String>]
    [-SkuName <String>]
    [-SkuSize <String>]
    [-SkuTier <String>]

```

With the flattened parameters, the new script would look like the following:

```powershell
New-AzResource `
    -Name storagesku `
    -ResourceGroupName skurg `
    -ProviderNamespace Microsoft.Storage `
    -ResourceType "storageAccounts" `
    -SkuName "Standard_LRS" `
    -SkuTier "Standard" `
    -Kind "Storage" `
    -ApiVersion "2019-04-01" `
    -Location centralus
```

> Some Network cmdlets use in-memory models or configurations for sub-resources that are then passed to creation cmdlets. To align all creation scenarios, we have removed all in-memory model and configuration creation commands. In `Az` 4.0 (Preview) and beyond you will create the all resources directly on Azure. A Subnet is a good example of this model as described in the [_Virtual Network Creation Quickstart_](https://docs.microsoft.com/en-us/azure/virtual-network/quick-create-powershell) document.

Reviewer: ```suggestion
Some Network cmdlets use in-memory models or configurations for sub-resources that are then passed to creation cmdlets. To align all creation scenarios, we have removed all in-memory model and configuration creation commands. In `Az` 4.0 (Preview) and beyond you will create all resources directly in Azure. A Subnet is a good example of this model as described in the [_Virtual Network Creation Quickstart_](https://docs.microsoft.com/en-us/azure/virtual-network/quick-create-powershell) document.
```

#### `Az` 2.6

```powershell
$agSubnetConfig = New-AzVirtualNetworkSubnetConfig `
    -Name myAGSubnet `
    -AddressPrefix 10.0.1.0/24
$backendSubnetConfig = New-AzVirtualNetworkSubnetConfig `
    -Name myBackendSubnet `
    -AddressPrefix 10.0.2.0/24
New-AzVirtualNetwork `
    -ResourceGroupName myResourceGroupAG `
    -Location eastus `
    -Name myVNet `
    -AddressPrefix 10.0.0.0/16 `
    -Subnet $agSubnetConfig, $backendSubnetConfig
```

With the new model of creating resources inline, you can replace the configuration step with an array of hashtable (_note_: see model flattening for more details). Here we have the property `AddressPrefix` to differentiate the VNet address prefix with subnet address prefix. Future releases can rename the flattened property here as subnet address prefix.

#### `Az` 4.0 (Preview)

```powershell
$agSubnetConfig = @{ Name= "myAGSubnet"; AddressPrefix = "10.0.1.0/24" }
$backendSubnetConfig = @{ Name= "myBackendSubnet"; AddressPrefix = "10.0.2.0/24" }
$vn1 = New-AzVirtualNetwork `
    -Name myvnet `
    -ResourceGroupName myrg `
    -location centralus `
    -AddressPrefix 10.0.0.0/16 `
    -Subnet $agSubnetConfig, $backendSubnetConfig

> ```

Reviewer: We need sections for additional comments, right?

$agSubnetConfig = New-AzVirtualNetworkSubnetConfig `
    -Name myAGSubnet `
    -AddressPrefix 10.0.1.0/24
$backendSubnetConfig = New-AzVirtualNetworkSubnetConfig `
    -Name myBackendSubnet `
    -AddressPrefix 10.0.2.0/24
New-AzVirtualNetwork `
    -ResourceGroupName myResourceGroupAG `
    -Location eastus `
    -Name myVNet `
    -AddressPrefix 10.0.0.0/16 `
    -Subnet $agSubnetConfig, $backendSubnetConfig
```

With the new model of creating resources inline, you can replace the configuration step with an array of hashtable (_note_: see model flattening for more details). Here we have the property `AddressPrefix` to differentiate the VNet address prefix with subnet address prefix. Future releases can rename the flattened property here as subnet address prefix.

#### `Az` 4.0 (Preview)

```powershell
$agSubnetConfig = @{ Name= "myAGSubnet"; AddressPrefix = "10.0.1.0/24" }
$backendSubnetConfig = @{ Name= "myBackendSubnet"; AddressPrefix = "10.0.2.0/24" }
$vn1 = New-AzVirtualNetwork `
    -Name myvnet `
    -ResourceGroupName myrg `
    -location centralus `
    -AddressPrefix 10.0.0.0/16 `
    -Subnet $agSubnetConfig, $backendSubnetConfig
```
## Comments and Questions

> Add here and for comments and questions about he proposed change

Reviewer: ```suggestion
Add your comments and questions about the proposed change
```
